File: plox/pylox/environment.py
import logging

from pylox.exceptions import RuntimeException


logger = logging.getLogger("pylox.environment")


class Environment(object):
    """Runtime environment which stores/updates named variables throughout the lifetime of a lox execution."""
    def __init__(self):
        self.values = {}

    def define(self, name, value):
        logger.debug("Defining var '{}' := {}".format(name, value))
        self.values[name] = value

    def get(self, name):
        if name.lexeme in self.values:
            return self.values[name.lexeme]
        raise RuntimeException(name.lexeme, "Undefined variable '{}'.".format(name.lexeme))

    def assign(self, name, value):
        if name.lexeme in self.values:
            self.values[name.lexeme] = value
            return
        raise RuntimeError("Undefined variable '{}'".format(name.lexeme))

    def __repr__(self):
        return "Environment({})".format(self.values)

File: klox/src/main/kotlin/environment.kt
package com.mab.lox

class Environment(val enclosing: Environment? = null) {
    private val values = HashMap<String, Any?>()

    fun define(name: String, value: Any?) {
        values[name] = value
    }

    fun get(name: Token): Any? {
        if (values.containsKey(name.lexeme)) {
            return values[name.lexeme]
        }

        if (enclosing != null) {
            return enclosing.get(name)
        }

        throw RuntimeError(name, "Undefined variable '${name.lexeme}'.")
    }

    fun assign(name: Token, value: Any?) {
        if (values.containsKey(name.lexeme)) {
            values[name.lexeme] = value
            return
        }

        if (enclosing != null) {
            enclosing.assign(name, value)
            return
        }

        throw RuntimeError(name, "Undefined variable '${name.lexeme}'.")
    }
}